lambda-buffers-compiler/src/LambdaBuffers/Compiler/TypeClassCheck/SuperclassCycleCheck.hs

```diff
@@ -0,0 +1,98 @@
+module LambdaBuffers.Compiler.TypeClassCheck.SuperclassCycleCheck (runCheck) where
+
+import Control.Lens ((&), (.~), (^.))
+import Control.Monad (when)
+import Control.Monad.Error.Class (MonadError (throwError))
+import Control.Monad.Except (runExcept)
+import Control.Monad.Reader (ReaderT (runReaderT))
+import Control.Monad.Reader.Class (MonadReader, asks, local)
+import Data.Foldable (Foldable (foldl', toList), for_)
+import Data.Map (Map)
+import Data.Map qualified as Map
+import Data.ProtoLens (Message (defMessage))
+import Data.Set (Set)
+import Data.Set qualified as Set
+import LambdaBuffers.Compiler.ProtoCompat qualified as PC
+import Proto.Compiler qualified as P
+import Proto.Compiler_Fields (unboundClassRefErr)
+import Proto.Compiler_Fields qualified as P
+
+data CheckRead = MkCheckRead
+  { currentModuleName :: PC.ModuleName
+  , currentClassName :: PC.ClassName
+  , classDefs :: Map (PC.InfoLess PC.ModuleName, PC.InfoLess PC.ClassName) PC.ClassDef
+  , trace :: Set (PC.InfoLess PC.ClassName)
+  , reportingTrace :: [PC.TyClassRef]
+  }
+
+type MonadCheck m = (MonadReader CheckRead m, MonadError P.TyClassCheckError m)
+
+runCheck :: PC.CompilerInput -> Either [P.TyClassCheckError] ()
+runCheck ci =
+  let -- Index the class definitions by a qualified name (ModuleName, ClassName)
+      classDefs =
+        Map.fromList
+          [ ((mn, cn), cd)
+          | (mn, m) <- Map.toList $ ci ^. #modules
+          , (cn, cd) <- Map.toList $ m ^. #classDefs
+          ]
+      -- Process each type class definitions under the appropriate reader context and collect errors
+      allErrors =
+        foldl'
+          ( \errs (mn, cd) ->
+              let errM = runReaderT (checkClassDef cd) (MkCheckRead mn (cd ^. #className) classDefs Set.empty [])
+               in case runExcept errM of
+                    Left err -> err : errs
+                    Right _ -> errs
+          )
+          []
+          [(m ^. #moduleName, cd) | m <- toList $ ci ^. #modules, cd <- toList $ m ^. #classDefs]
+   in if null allErrors then Right () else Left allErrors
```

Comment:
This implementation is very hard to understand. As you're accumulating some errors and reporting if you found any refactor this to use a `WriterT [Err] Reader ...`  + traverse.

Reply:
Good idea! On it!

Reply:
Ok I tried but then realized that actually there's a reason I did it like this.

I can't just accumulate errors in Writer, because once the cycle error has been found it has to terminate the check for that particular classdef. So on each error I really do have to terminate the Error computation and throwError rather than tell.

Reply:
I tried refactoring a bit to make it a bit more comprehensible.